### PR TITLE
Add New T2iAdapter Training script [Light Control]

### PR DESCRIPTION
**Pull request by nbardy**
# What does this PR do?

This branch add a training script that uses a pre-processing step to generate light and color signals to condition the model on.

Allows conditioning on user generated in code signals instead of only pre-existing data sets

[Currently Drafting] 

Initial model results promising

### CODE REVIEW

**Comment by nbardy** [comment]
Training a model with this + v-diffusion today to see if there is any noticeable difference.

**Review by williamberman — APPROVED** [review]
tbh I'm not super familiar with scheduler snr stuff so someone please confirm that this change is sound. But code wise looks fine

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 1]
```diff
@@ -408,6 +408,12 @@ def parse_args(input_args=None):
         default=4,
         help=("The dimension of the LoRA update matrices."),
     )
+    parser.add_argument(
+        "--scale_scheduler",
+        action="store_true",
+        default=False,
+        help="Rescale Scheduler based on Zero SNR paper during training and sampling",
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:415]: Let's also add the link for reference here.

File: examples/text_to_image/train_text_to_image_lora_sdxl.py [hunk 2]
```diff
@@ -1228,6 +1241,7 @@ def compute_time_ids(original_size, crops_coords_top_left):
         pipeline = StableDiffusionXLPipeline.from_pretrained(
             args.pretrained_model_name_or_path, vae=vae, revision=args.revision, torch_dtype=weight_dtype
         )
+        pipeline.scheduler = noise_scheduler
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora_sdxl.py:1244]: I would keep the default noise scheduler and instead initialize it have the additional arguments actually if that's okay. Here's an example:

https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora.py#L1362C1-L1372C114

Here, instead we would do:

```python
pipeline.scheduler = pipeline.scheduler.from_config(pipeline.scheduler.config, **scheduler_args)
```

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 3]
```diff
@@ -698,7 +704,13 @@ def main(args):
...
+    noise_scheduler = DDPMScheduler.from_pretrained(
+        args.pretrained_model_name_or_path,
+        subfolder="scheduler",
+        rescale_betas_zero_snr=args.scale_scheduler,
+        timestep_spacing="trailing" if args.scale_scheduler else None,
+    )
+
     text_encoder_one = text_encoder_cls_one.from_pretrained(
         args.pretrained_model_name_or_path, subfolder="text_encoder", revision=args.revision
     )
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:716]: I think we also need to incorporate https://github.com/huggingface/diffusers/pull/4623/files#r1295368965 here in this script too?

File: examples/text_to_image/train_text_to_image_lora_sdxl.py [hunk 4]
```diff
@@ -560,7 +566,12 @@ def main(args):
     )
 
     # Load scheduler and models
-    noise_scheduler = DDPMScheduler.from_pretrained(args.pretrained_model_name_or_path, subfolder="scheduler")
+    noise_scheduler = DDPMScheduler.from_pretrained(
+        args.pretrained_model_name_or_path,
+        subfolder="scheduler",
+        rescale_betas_zero_snr=args.scale_scheduler,
+        timestep_spacing="trailing" if args.scale_scheduler else None,
+    )
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora_sdxl.py:574]: If it's used with the Min-SNR strategy, this might lead to numerical instabilities: 

https://github.com/huggingface/diffusers/blob/a7de96505b8c71a626c6fa45ff3cf85e42db4960/examples/text_to_image/train_text_to_image_lora_sdxl.py#L666

This is how we deal with it [here](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_sdxl.py#L557C1-L568C10). Cc: @bghira.

**Review by sayakpaul — COMMENTED** [review]
Looking amazing. Just left a few comments. 

Could you additionally post some results comparison with and without Zero SNR?

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 5]
```diff
@@ -698,7 +704,13 @@ def main(args):
...
 
     # Load scheduler and models
-    noise_scheduler = DDPMScheduler.from_pretrained(args.pretrained_model_name_or_path, subfolder="scheduler")
+    noise_scheduler = DDPMScheduler.from_pretrained(
+        args.pretrained_model_name_or_path,
+        subfolder="scheduler",
+        rescale_betas_zero_snr=args.scale_scheduler,
+        timestep_spacing="trailing" if args.scale_scheduler else None,
```
Comment (bghira) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:711]: DDPM doesn't support these

File: examples/text_to_image/train_text_to_image_lora_sdxl.py (hunk 2 above)
Comment (bghira) [reply examples/text_to_image/train_text_to_image_lora_sdxl.py:1244]: reusing the noise scheduler from training also didn't work here.

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 1 above)
Comment (bghira) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:415]: i would prefer the argument be `--rescale_betas_zero_snr`

File: examples/dreambooth/train_dreambooth_lora_sdxl.py (hunk 5 above)
Comment (bghira) [reply examples/dreambooth/train_dreambooth_lora_sdxl.py:711]: also not sure whether we want to unconditionally override the training timestep spacing, because the original paper also shows that it works with `leading` timestep spacing.


it's not clear whether the timestep spacing matters at training time.

File: examples/text_to_image/train_text_to_image_lora_sdxl.py [hunk 6]
```diff
@@ -417,6 +417,12 @@ def parse_args(input_args=None):
         default=4,
         help=("The dimension of the LoRA update matrices."),
     )
+    parser.add_argument(
+        "--scale_scheduler",
+        action="store_true",
+        default=False,
+        help="Rescale Scheduler based on Zero SNR paper during training and sampling",
```
Comment (bghira) [inline examples/text_to_image/train_text_to_image_lora_sdxl.py:424]: there should probably be some mention about how terminal SNR requires a v_prediction model, as predicting epsilon results in no meaningful learning by the model at training time. this is stated in the original paper.

it feels like having a LoRA that can rescale betas on its own has zero value, we should be simply following the pretrained base model scheduler config.

**Comment by bghira** [comment]
code-wise there are some issues, we can't set these directly in DDPM. that is required as part of a larger set of changes.

currently you have to create a reference scheduler via DDIM and grab the betas from there, and pass them into DDPM as pretrained_betas.

additionally, I have been tuning SDXL on v_prediction for a week. it doesn't seem to be working.

there's either something wrong with:

* rescaling of betas 
* rescaling of timesteps
* rescaling of CFG

image

It's honestly very distressing. I would recommend that a deeper dive into the issue with @PeterL1n be done first.

**Comment by bghira** [comment]
@nbardy you might have accidentally pushed a t2i script here.

**Comment by nbardy** [comment]
Oops. Yes. This is based off my main branch

I never did come back and do the cleanup on this branch to merge it. Maybe I should just close this PR.

**Comment by nbardy** [comment]
I shouldn't of made this PR with my main branch. It's become super tangled up now in the commits. Going to move this to draft.

I can split out the Zero-SNR stuff separate.

**Comment by bghira** [comment]
can you split out the two branches?

**Comment by bghira** [comment]
this pull request is going off the rails. you are adding so many things that do not exist in other portions of Diffusers, including lion optimizer now from an external project.

* can you split the PRs out into two, as you initially meant to?
* can you reduce reliance on external dependencies?

**Comment by nbardy** [comment]
Should of just closed it not went to draft. Did not mean to propose all to upstream this is experimenting Im running on my fork. Had opened a PR original for a few commits that was mistakenly pointed at upstream main.
